esphome/components/wifi_info/wifi_info_text_sensor.cpp

```diff
@@ -2,18 +2,121 @@
 #ifdef USE_WIFI
 #include "esphome/core/log.h"
 
-namespace esphome {
-namespace wifi_info {
+namespace esphome::wifi_info {
 
 static const char *const TAG = "wifi_info";
 
+static constexpr size_t MAX_STATE_LENGTH = 255;
+
+/********************
+ * IPAddressWiFiInfo
+ *******************/
+
+void IPAddressWiFiInfo::setup() {
+  wifi::global_wifi_component->add_on_ip_state_callback(
+      [this](const network::IPAddresses &ips, const network::IPAddress &dns1_ip, const network::IPAddress &dns2_ip) {
+        this->state_callback_(ips);
+      });
+}
+
 void IPAddressWiFiInfo::dump_config() { LOG_TEXT_SENSOR("", "IP Address", this); }
+
+void IPAddressWiFiInfo::state_callback_(const network::IPAddresses &ips) {
+  this->publish_state(ips[0].str());
+  uint8_t sensor = 0;
+  for (auto &ip : ips) {
```

Comment:
```suggestion
  for (const auto &ip : ips) {
```

nit: ip is never modified. Will make no difference though since compiler will know